subsys/net/lib/sockets/sockets.c

```diff
@@ -1572,21 +1580,28 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 	if (msg != NULL) {
 		if (msg->msg_control != NULL) {
 			if (msg->msg_controllen > 0) {
-				bool clear_controllen = true;
+				bool clear_controllen = false;
```

Comment:
after some more thought, I think this `clear_controllen` logic is not working well.
If we have both TIMESTAMPING and RECV_PKTINFO enabled, we might end up in some corner cases when we clear `msg_controllen` even if we shouldn't.
The only safe way to implement this seems to compute the control data length in this function